### PR TITLE
packaging + environment reproducibility (cuda, cpu, mps) + bugfixes

### PR DESCRIPTION
**Pull request by mathematicalmichael**
Hello, thanks for this repo! I'm super excited to start playing around with your latest work.

I encountered some issues that I was able to resolve with minimal effort, and want to contribute them back.

I wanted to run a basic test script:

```python
from parampacmap import ParamPaCMAP

import numpy as np

A = np.random.randn(100, 20)

r = ParamPaCMAP().fit_transform(A)

print(r)
```

without a GPU, in a Github Codespace.

```bash
uv run --with-editable '.[cpu]' tests/test_script.py
```
so this PR includes the fixes for that, which involved:
- missing arguments
- type casting for numba

it also specifies its torch dependencies with `[cpu]` vs various cuda-based ones.

this could use a lot more testing with different environments, but I'd like to propose this PR as a start.

UPDATE:
this PR has been tested against:
- cpu
- three cuda torch versions
- mps (metal)

### CODE REVIEW

**Comment by mathematicalmichael** [comment]
on `Ubuntu 20.04.5 LTS focal x86_64 Kernel: 5.15.0-122-generic` with cuda 12.2 in a docker container:

```bash
NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2 
```

```bash
uv run --with-editable '.[cu121]' test_script.py
```
works! (after `apt install clang`)

as does `uv run --with-editable '.[cu118]' test_script.py` (older kernel still supported)

`uv run --with-editable '.[cu124]' test_script.py` also works...

(i did verify that cuda was being used)

**Comment by mathematicalmichael** [comment]
I got m1 macOS support working by adding an `if __name__ == "__main__":` guard, and included the test file under `tests/`.

also got it working with `pytest` as a basic starting point example.

still struggling with specifying `torch` as a core dependency given the mix of OS-specific indices. I am testing on three platforms to be sure there's a path for reproducibility on all three at the very least.

```bash
pip install -e .[cpu]  # for cpu-only envs
pip install -e .[mps]  # for M-series macs, aarch64
pip install -e .[cu121] # cuda ... three of these - cu118, cu121, cu124, reflecting torch's current releases.
```

**Comment by mathematicalmichael** [comment]
tested this on the mammoth dataset from the pacmap repo and found the results look right with the default class instantiated with 2 dimensions. csv data + labels from "umap version" json file:

image

**Comment by mathematicalmichael** [comment]
I am, however, finding that `seed` is not leading to deterministic results, as is expectation.

File: README.md [hunk 1]
```diff
@@ -11,6 +11,26 @@ cd ParamRepulsor
...
+
+```bash
+pip install .[cpu]    # cpu-only pytorch
+pip install .[cu118]  # cuda 118
+pip install .[cu121]  # cuda 121
+pip install .[cu124]  # cuda 124
+pip install .[mps]    # arm64/aarch64 (Apple M-Series chips)
+```
+
+This project also supports `uv` (`pip install uv`):
```
Comment (hyhuang00) [inline README.md:25]: Haven't tried uv before, this looks very interesting!

**Comment by hyhuang00** [comment]
Hi Michael,

Thank you so much for your valuable contributions to improving the ParamRepulsor repository. We’ve been reviewing the repository structure and noticed that including author references in the pyproject.toml file isn’t a standard practice in open-source projects. It might also cause some confusion, given that this repository primarily serves as the codebase for our paper.

That said, we absolutely want to ensure your contributions are properly acknowledged. To address this, we plan to create a dedicated contributors file and reference it in the README. This way, your efforts will be clearly recognized. Thank you again for your understanding and collaboration.

Best,
Haiyang

**Comment by mathematicalmichael** [comment]
@hyhuang00 that's fine, I'll make the update.
The migration to pyproject.toml-only is relatively new, and as I understand it all the metadata you put in there shows up on pypi:
image
to assist in finding contact info for the author.

So, do you want me to remove the section altogether or just my name?

**Review by hyhuang00 — APPROVED** [review]
LGTM

**Comment by hyhuang00** [comment]
Again, thank you so much for your help!
